src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAuthenticationManager.java

```diff
@@ -345,28 +330,66 @@ public RegisteredUser getUserFromSession(final HttpServletRequest request) {
             }
         }
 
-        // check if the users session is valid.
-        if (!this.isValidUsersSession(currentSessionInformation)) {
-            log.debug("User session has failed validation. Assume they are not logged in. Session: "
-                    + currentSessionInformation);
+        // retrieve the user from database.
+        try {
+            // Get the user the cookie claims to belong to from the session information:
+            long currentUserId = Long.parseLong(currentSessionInformation.get(SESSION_USER_ID));
+            RegisteredUser userToReturn =  database.getById(currentUserId);
+
+            // Check that the user's session is indeed valid:
+            if (!this.isValidUsersSession(currentSessionInformation, userToReturn)) {
+                log.debug("User session has failed validation. Treating as logged out. Session: " + currentSessionInformation);
+                return null;
+            }
+            
+            return userToReturn;
+        } catch (SegueDatabaseException e) {
+            log.error("Internal Database error. Failed to resolve current user.", e);
+            return null;
+        } catch (NumberFormatException e) {
+            log.info("Invalid user id detected in session. " + currentSessionInformation.get(SESSION_USER_ID));
+            return null;            
+        }
+    }
+
+    /**
+     * @see #getUserFromSession(HttpServletRequest) - the two types of "request" have identical methods but are not
```

Comment:
Not sure I understand the need to duplicate methods in this way? Could it be refactored?

Reply:
I'd love for it to be, but they're different classes with different class structures and return types for "getCookies", so I can't find a way to make them use the same method.

Reply:
I'm glad to see that single inheritance is causing more problems than it is solving 👍